PSA Initial Attestation service

moranpeker opened the pull request:
### Description

Attestation service can create a token on request, which contains a fix set of
device specific data.

Contains the implementation of ‘psa_initial_attest_get_token_size’ , ‘psa_initial_attest_get_token’ and ‘psa_attestation_inject_key’.

PSA Initial Attestation Service based on TF-M Attestation Service that taken from TF-M master.
TF-M latest attestation version is: c43181daf54f69f53de58593a50dd6a9c233eecd

_**Signing of token**_
‘psa_attestation_inject_key’ API - Device must contain an attestation key pair. The token is signed with the private part of attestation key pair. The public key is used to verify the token authenticity.

Inject attestation key by importing a given private key or generate a key pair.
Attestation key stored as a persistent key with a specific key-id.

_**Claims in the initial attestation token**_
The initial attestation token is formed of claims.
List of claims are included in the token:
https://confluence.arm.com/display/IoTBU/Claims+in+PSA+IAT+report
In the current implementation no bootloader exists in single & dual V7 .
In the current implementation temporary mandatory claims (ONLY) are given by a hard coded temp bootloader status data (attestation_bootloader_data.c).

_**Code structure**_
The PSA interface for the Initial Attestation Service is located in
` components\TARGET_PSA\services\attestation `.
The only headers to be included by applications that want to use functions from
the PSA API are `psa_initial_attestation_api.h` and `psa_attest_inject_key.h` (under ` components\TARGET_PSA\inc\psa folder`).

**_Service source files_**
- CBOR library: `qcbor`: This library is used to create a proper CBOR token. It is a  fork of this external  [QCBOR library](https://github.com/laurencelundblade/QCBOR).

- For now, TFM source files are part of this PR and located in ` components\TARGET_PSA\services\attestation\tfm_impl ` +  header files ‘psa_initial_attestation_api.h’ and ‘attestation.h’.
An importer needs to be created to import TFM source files.

Code checked over K64F and PSOC6.
Pass compliance tests over  K64F and PSOC6.

This PR includes and depends on the security lifecycle bug fix ( PR = #9745 ) 

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [X] Breaking change

### Reviewers

<!-- 
    Optional
    Request additional reviewers with @username
-->

### Release Notes

* _A brief description of changes introduced._
  Add PSA Initial Attestation Service implementation.
* _An analysis of effects: components affected, potential consequences for users and reasons for the addition or change._
this code is a new feature and doesn't effect existing code.
* _Migration guidance: actions for updating the current code. Please include code snippets to illustrate before and after the addition or change._
 This breaks backwards compatibility with Nucleo F411RE entropy injection.



<!--
   Optional
   In case of breaking changes, functionality changes or refactors, please add release notes here.
   For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

Commit:
Add PSA Initial Attestation service

Attestation service can create a token on request, which contains a fix set of
device specific data.
Implementation:
-‘psa_initial_attest_get_token_size’- get exact size of initial attestation token in bytes.
-‘psa_initial_attest_get_token’- get the initial attestation token.
-‘psa_attestation_inject_key’ - Generate or import the attestation key pair and export the public part.

-Including CBOR lib and TFM attestation implemantation.
-Temporary claim’s data – no bootloader over V7 Single & Dual

Commit:
Attestation: aut- generated files by generate_partition_code.py

Atestation paration added.

Commit:
Attestation: Update astyle.ignore

Ignore TFM imported code from astyle

Commit:
Attestation: update targets.json for K64F

Make K64F run with MBED_SPM
Attestation code using SPM structures over single and dual core.

Commit:
Attestation: Add test for initial attestation service

Add single test that check inject attestation key and attestation get token

ciarmcom commented:
@moranpeker, thank you for your changes.
@ARMmbed/mbed-os-maintainers please review.

cmonr commented on .astyleignore:
```diff
@@ -23,3 +23,4 @@ targets
 components/802.15.4_RF
 components/wifi
 tools
+components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/tfm_impl
```
👍 for the commit comment about this.

0xc0170 replied:
One note: please revert file permissions changes `100644 → 100755` ?

cmonr commented:
> Checked over K64F and PSOC6.

Should PSOC6 also have its/their targets.json entries updated?

cmonr reviewed (dismissed):
A nit whilse others get a better chance to review.
Please revert the permissions of target.json and .astyle.yml (755 -> 644)

cmonr commented:
Also, there are still some astyle issues: https://travis-ci.org/ARMmbed/mbed-os/jobs/491854379

Commit:
Atteststaion: Update header files

-Remove psa_initial_attesttaion.h
-Move psa_initial_attesttaion_api.h and psa_attest_inject_key.h into
inc folder
-Update include in test file

Commit:
Attestation: astyle fixs

Runing astyle over attestation file to pass CI

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_EMUL/psa_attest_inject_key.c:
```diff
@@ -0,0 +1,26 @@
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
```
please change license for all to Apache.
 * Copyright (c) 2006-2013 ARM Limited
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

alekshex replied:
2018-2019

alekshex replied:
Please leave all TF-M sources license as is. only our wrappers should be under Apache.
TF-M license is handled by legal currently

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_EMUL/psa_initial_attestation_api.c:
```diff
@@ -0,0 +1,63 @@
+        return err;
+    }
+
+    *token_size = out_vec[0].len;
```
do we need this line?
or this takes care of it psa_outvec out_vec[1] = { { token, *token_size } };

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_boot_status_loader.c:
```diff
@@ -0,0 +1,76 @@
+ * \brief Indicates whether shared data area was already initialized.
+ *
+ */
+static uint32_t shared_data_init_done;
```
please assign initial uninitialized value

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_EMUL/psa_initial_attestation_api.c:
```diff
@@ -0,0 +1,63 @@
+    psa_outvec out_vec[1] = { { token_size, sizeof(*token_size) } };
+
+    err = initial_attest_get_token_size(in_vec, 1, out_vec, 1);
+    if (err != PSA_ATTEST_ERR_SUCCESS)
```
no need, err returned anyway, unless you want additional handling

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_boot_status_loader.c:
```diff
@@ -0,0 +1,76 @@
+    offset  = (uintptr_t)BOOT_TFM_SHARED_DATA_BASE + (uintptr_t)SHARED_DATA_HEADER_SIZE;
+
+    /* Add header to output buffer as well */
+    if (len < SHARED_DATA_HEADER_SIZE)
```
what if len > header size?
is the idea that len must be equal?

moranpeker replied:
len its the buffer lenght. it must be bigger/equal to all boot status data.
this check in line 78:
```if (len < ptr_tlv_header->tlv_tot_len + tlv_entry->tlv_len)```

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_boot_status_loader.c:
```diff
@@ -0,0 +1,76 @@
+    if (len < SHARED_DATA_HEADER_SIZE)
+    {
+        return PSA_ATTEST_ERR_INIT_FAILED;
+    } else
```
why do you need else if you have return before?

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,186 @@
+#include "tfm_plat_crypto_keys.h"
+#include <string.h>
+
+static psa_hash_operation_t hash_handle;
```
please init to NULL initially

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,186 @@
+
+    (void)key_select;
+
+    const psa_key_id_t key_id = 17;
```
17, please move to define with meaninful name.
something like PSA_ATTESTATION_PRIVATE_KEY_ID
or so

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,186 @@
+    crypto_ret = psa_open_key(PSA_KEY_LIFETIME_PERSISTENT, key_id, &handle);
+    if (crypto_ret != PSA_SUCCESS)
+    {
+        return T_COSE_ERR_NO_KID;
```
should we cleanup init if we fail?
mbedtls_psa_crypto_free( );?

moranpeker replied:
I removed all calling to ```psa_crypto_init```  - we assume init called before crypto operations, if not an error will be return.

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,186 @@
+    enum tfm_plat_err_t err;
+    enum ecc_curve_t cose_curve;
+    struct ecc_key_t attest_key = {0};
+    uint8_t  key_buf[ECC_P_256_KEY_SIZE];
```
uint8_t  key_buf[ECC_P_256_KEY_SIZE]  = 0;

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,186 @@
+
+    crypto_ret = psa_hash_setup(&hash_handle, PSA_ALG_SHA_256);
+
+    if (crypto_ret == PSA_ERROR_NOT_SUPPORTED)
```
just a proposal
you can do error translation function for all psa to cose errors.
it will accept psa return code (even success) and will map it to relevant cose error.
then you will save all those if else statements throughout the code

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,186 @@
+    if (data_to_hash.ptr != NULL)
+    {
+        psa_hash_update(&hash_handle, data_to_hash.ptr, data_to_hash.len);
+    } else
```
why you need else then?

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto_keys.c:
```diff
@@ -0,0 +1,186 @@
+    uint32_t initial_attestation_public_x_key_size = 0;
+    uint32_t initial_attestation_public_y_key_size = 0;
+
+    const psa_key_id_t key_id = 17;
```
please see comment above.

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto_keys.c:
```diff
@@ -0,0 +1,186 @@
+    crypto_ret = psa_open_key(PSA_KEY_LIFETIME_PERSISTENT, key_id, &handle);
+    if (crypto_ret != PSA_SUCCESS)
+    {
+        return TFM_PLAT_ERR_SYSTEM_ERR;
```
no crypto deinit?

moranpeker replied:
I removed all calling to `psa_crypto_init` - we assume init called before crypto operations, if not an error will be return.

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_iat_claims_loader.c:
```diff
@@ -0,0 +1,176 @@
+#include "attestation.h"
+#include "crypto.h"
+
+extern int32_t g_caller_id;
```
must be initialized

moranpeker replied:
no need in extern

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_iat_claims_loader.c:
```diff
@@ -0,0 +1,176 @@
+/* Hash of attestation public key */
+static enum tfm_plat_err_t attest_public_key_sha256(uint32_t *size, uint8_t *buf)
+{
+    const psa_key_id_t key_id = 17;
```
...

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_iat_claims_loader.c:
```diff
@@ -0,0 +1,176 @@
+
+    crypto_ret = psa_open_key(PSA_KEY_LIFETIME_PERSISTENT, key_id, &handle);
+    if (crypto_ret != PSA_SUCCESS) {
+        return TFM_PLAT_ERR_SYSTEM_ERR;
```
..

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/psa_inject_attestation_key_impl.c:
```diff
@@ -0,0 +1,99 @@
+
+    status = psa_create_key(lifetime, key_id, &handle);
+    if (status != PSA_SUCCESS) {
+        return (status);
```
deinit

moranpeker replied:
same.

alekshex reviewed (commented):
first pass

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,184 @@
+{
+    if (data_to_hash.ptr != NULL) {
+        psa_hash_update(&hash_handle, data_to_hash.ptr, data_to_hash.len);
+    } else {
```
why else needed?

moranpeker replied:
This function caled by get attestation token and get attestation token size. 
for the second purpose - data_to_hash(the payload) its null and no hash needed

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IPC/psa_attest_inject_key.c:
```diff
@@ -0,0 +1,55 @@
+        public_key_data, public_key_data_size
+    };
+    out_vec[1] = (psa_outvec) {
+        public_key_data_length, sizeof(*public_key_data_length)
```
why sizeof?

moranpeker replied:
why not ?

alekshex replied:
is the idea here to have sizeof member pointed to?

moranpeker replied:
yes

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IPC/psa_initial_attestation_api.c:
```diff
@@ -0,0 +1,72 @@
+    psa_outvec out_vec[1] = { { token, *token_size } };
+
+    handle = psa_connect(PSA_ATTEST_GET_TOKEN_ID, MINOR_VER);
+    if (handle <= 0)
```
0 is also illegal?

moranpeker replied:
right

Commit:
Attestation: Change files license to Apache

change all files except tfm code.

Commit:
Attestation: Add a new define for attestation private key id

Commit:
Attestation: Update iat claims data

Add more iat claims: profile definition and verification service
Update boot status data: add more SW components for SW_S_NS module

Commit:
Add checks ,initialization,remove unneeded code and files license update

0xc0170 commented on TESTS/psa/attestation/main.cpp:
```diff
@@ -0,0 +1,148 @@
+/*
+* Copyright (c) 2018 ARM Limited. All rights reserved.
```
2019 now?

0xc0170 reviewed (dismissed):
> [ ] Functionality change

Please set PR type

If functionality change, I believe this one yes, will require release notes addition , see https://os.mbed.com/docs/mbed-os/v5.11/contributing/workflow.html#functionality-change

0xc0170 commented on TESTS/psa/attestation/main.cpp:
```diff
@@ -0,0 +1,148 @@
+    psa_key_handle_t handle = 0;
+    psa_open_key(PSA_KEY_LIFETIME_PERSISTENT, key_id, &handle);
+    psa_destroy_key(handle);
+    // mbedtls_psa_cr/ypto_free();
```
don't leave dead code behind

0xc0170 commented on components/TARGET_PSA/inc/psa/psa_attest_inject_key.h:
```diff
@@ -0,0 +1,88 @@
+/*   These APIs are still evolving and are meant as a prototype for review.*/
+/*   The APIs will change depending on feedback and will be firmed up      */
+/*   to a stable set of APIs once all the feedback has been considered.    */
+/***************************************************************************/
```
Is this ready fro the release or kept here only during this review?

0xc0170 commented on targets/targets.json:
```diff
@@ -1456,7 +1456,8 @@
             "MCU_K64F",
             "Freescale_EMAC",
-            "PSA"
+            "PSA",
```
Same for this file, file permissions are being changed

0xc0170 commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,230 @@
+// ---------------------------------- Includes ---------------------------------
```
missing license header

Commit:
Get security lifecycle by psa_security_lifecycle_state function

-Add bug fix for psa_security_lifecycle_state function - PR #9745
-Remove lifecycle data from temp boot status data
-Update attesttation_partition_json file to use
PSA_PLATFORM_LC_GET and generate partition
code - auto generated files update

Commit:
Update to last version of TFM-Attesttaion code

-update TFM attestation files
-update attest_crypto.c and psa_attestation_stubs.c file to support TFM
code changes

Commit:
Update attestation asymmetric sign to use deterministic alg

- Add PSA_KEY_USAGE_VERIFY to attesttaion key usage
- Set deterministic alg to attestation key policy
- Call asymmetric sign with deterministic alg

Commit:
Revert file permissions changes 100644 to 100755

Commit:
Attestation: Add parameter and entry size checks

Commit:
Attestation: update boot status data

Epoc data to be 2 bytes instead of 4

Commit:
Attestation: Add close key after open key

Commit:
Attestation: move qcbor folder place

Commit:
Attestation: move t_code folder place

Commit:
Update astyleignore to ignore qcbor lib folder

Commit:
Update include psa_defs -> psa/client

Commit:
Add ifdef TARGET_PSA and MBEDTLS_PSA_CRYPTO_C to test file

if not supported return error

Commit:
Attestation: Add comments from temporary data

Commit:
generated file update

alekshex commented on TESTS/psa/attestation/main.cpp:
```diff
@@ -0,0 +1,151 @@
+{
+#if (defined(COMPONENT_PSA_SRV_IPC) || defined(MBEDTLS_ENTROPY_NV_SEED))
+    uint8_t seed[MBEDTLS_PSA_INJECT_ENTROPY_MIN_SIZE] = {0};
+    /* inject some a seed for test*/
```
/* inject some seed for test*/

alekshex commented on TESTS/psa/attestation/main.cpp:
```diff
@@ -0,0 +1,151 @@
+        seed[i] = i;
+    }
+
+    /* don't really care if this succeed this is just to make crypto init pass*/
```
/* don't really care if this succeeds this is just to make crypto init pass*/

alekshex commented on TESTS/psa/spm_client/COMPONENT_SPE/psa_setup.c:
```diff
@@ -28,14 +28,16 @@
 extern const uint32_t platform_external_sids[1];
 
-spm_partition_t g_partitions[4] = {
+spm_partition_t g_partitions[5] = {
```
why not in define?

moranpeker replied:
psa_setup is generated file by calling 'tools/psa/generate_tfm_partition_code.py' script

kjbracey replied:
That script could just omit the size, if only to avoid such comments.

alekshex commented on TESTS/psa/spm_server/COMPONENT_SPE/psa_setup.c:
```diff
@@ -29,15 +29,17 @@
 extern const uint32_t platform_external_sids[1];
 
-spm_partition_t g_partitions[5] = {
+spm_partition_t g_partitions[6] = {
```
def

moranpeker replied:
psa_setup is generated file by calling 'tools/psa/generate_tfm_partition_code.py' script

alekshex commented on TESTS/psa/spm_smoke/COMPONENT_SPE/psa_setup.c:
```diff
@@ -28,14 +28,16 @@
 extern const uint32_t platform_external_sids[1];
 
-spm_partition_t g_partitions[4] = {
+spm_partition_t g_partitions[5] = {
```
def

moranpeker replied:
psa_setup is generated file by calling 'tools/psa/generate_tfm_partition_code.py' script

alekshex commented on TESTS/psa/spm_smoke/COMPONENT_SPE/psa_setup.c:
```diff
@@ -102,11 +116,12 @@ uint32_t init_partitions(spm_partition_t **partitions)
 
     *partitions = g_partitions;
-    return 4;
+    return 5;
```
def

moranpeker replied:
psa_setup is generated file by calling 'tools/psa/generate_tfm_partition_code.py' script

alekshex commented on components/TARGET_PSA/TARGET_MBED_SPM/COMPONENT_SPE/psa_setup.c:
```diff
@@ -27,15 +27,28 @@
 
 __attribute__((weak))
-spm_partition_t g_partitions[3] = {
+spm_partition_t g_partitions[4] = {
```
def

moranpeker replied:
psa_setup is generated file by calling 'tools/psa/generate_tfm_partition_code.py' script

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_boot_status_loader.c:
```diff
@@ -0,0 +1,92 @@
+
+    shared_data_init_done = SHARED_DATA_INITIALZED;
+    return PSA_ATTEST_ERR_SUCCESS;
+}
```
is this legal EOF?

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_boot_status_loader.c:
```diff
@@ -0,0 +1,92 @@
+            memcpy(ptr, (const void *)tlv_entry, tlv_entry->tlv_len);
+            ptr += tlv_entry->tlv_len;
+            ptr_tlv_header->tlv_tot_len += tlv_entry->tlv_len;
+            if (len < ptr_tlv_header->tlv_tot_len) {
```
we may save the memcpy operation and be a bit safer in terms of security if we move the if before memcpy and do:
f (len < ptr_tlv_header->tlv_tot_len + tlv_entry->tlv_len)
  return....

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,285 @@
+     * failure.
+     */
+    cose_ret = check_hash_sizes();
+    if (cose_ret)
```
!= T_COSE_SUCCESS ?

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,285 @@
+            psa_hash_ctx->status = psa_hash_update(&psa_hash_ctx->operation,
+                                                   data_to_hash.ptr,
+                                                   data_to_hash.len);
+        } else {
```
why needed?

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,285 @@
+
+    if (psa_hash_ctx->status == PSA_SUCCESS) {
+        if (data_to_hash.ptr != NULL) {
+            psa_hash_ctx->status = psa_hash_update(&psa_hash_ctx->operation,
```
can update fail, should the return value be checked?

moranpeker replied:
if update fail you will return error while calling finish : 

https://github.com/ARMmbed/mbed-os/blob/1808bf26fe82033cf1134001debac8e203ed7cc4/components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c#L262-L282

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto_keys.c:
```diff
@@ -0,0 +1,195 @@
+    {
+        key_src = key_src + ONE_BYTE;
+        key_size = initial_attestation_public_x_key_size;
+        copy_key(key_dst, key_src, key_size);
```
how do you know that dst has enough space to hold src?

moranpeker replied:
mmm .. 
https://github.com/ARMmbed/mbed-os/blob/1808bf26fe82033cf1134001debac8e203ed7cc4/components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c#L104

ECC_P_256_KEY_SIZE  (96u)  ( 3 x 32 = 96 bytes priv + pub-x + pub-y )

key_dst --> key_buf
key_src --> the public key from ECC private attestation key ( )

public_key allready pass the check before to verify his type and size.
https://github.com/ARMmbed/mbed-os/blob/1808bf26fe82033cf1134001debac8e203ed7cc4/components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto_keys.c#L111-L139

I can had check if you like but I don't see an additional value.

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto_keys.c:
```diff
@@ -0,0 +1,195 @@
+    {
+        key_src += initial_attestation_public_x_key_size;
+        key_size = initial_attestation_public_y_key_size;
+        copy_key(key_dst, key_src, key_size);
```
how do you know that dst has enough space to hold src?

moranpeker replied:
same as comment before.

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_iat_claims_loader.c:
```diff
@@ -0,0 +1,228 @@
+                                 *size,
+                                 (size_t *) size);
+    if (crypto_ret != PSA_SUCCESS) {
+        free(public_key);
```
too many same kind of operations: free, close
please consider goto with appropriate error code

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_iat_claims_loader.c:
```diff
@@ -0,0 +1,228 @@
+ * \param[in]   p_src  Pointer to the ID
+ * \param[in]   size   Length of the ID
+ */
+static inline void copy_id(uint8_t *p_dst, uint8_t *p_src, size_t size)
```
code duplication, copy_key can be used here also.
just name it copy_src_to_dst

And if the ptr is uint8*, why aren't you using memcpy??

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attestation_bootloader_data.c:
```diff
@@ -0,0 +1,49 @@
+
+#include "attestation_bootloader_data.h"
+
+/* Temporary Boodloader data - conatians temp mandatory claims */
```
contains

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/psa_attestation_stubs.c:
```diff
@@ -0,0 +1,31 @@
+        return PSA_ATTEST_ERR_INVALID_INPUT;
+    }
+    return PSA_ATTEST_ERR_SUCCESS;
+}
```
legal eof?

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/psa_inject_attestation_key_impl.c:
```diff
@@ -0,0 +1,110 @@
+
+    status = psa_create_key(lifetime, key_id, &handle);
+    if (status != PSA_SUCCESS) {
+        return (status);
```
no crypto deinit?

moranpeker replied:
same

alekshex reviewed (commented):
please see comments

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,248 @@
+                SPM_PANIC("SPM read length mismatch");
+            }
+
+            token_buff = calloc(1, msg.out_size[0]);
```
where is this being freed?

moranpeker replied:
fixed

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,248 @@
+
+            token_buff = calloc(1, msg.out_size[0]);
+            if (token_buff == NULL) {
+                status = PSA_ATTEST_ERR_GENERAL;
```
free(challenge_buff);

alekshex reviewed (commented):
please see regarding memory management

alekshex commented on components/TARGET_PSA/services/platform/COMPONENT_PSA_SRV_EMUL/platform_emul.c:
```diff
@@ -21,7 +21,12 @@
 uint32_t psa_security_lifecycle_state(void)
```
license header?

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,248 @@
+            set_caller_id(msg);
+            status = initial_attest_get_token(in_vec, 1, out_vec, 1);
+            if (status == PSA_ATTEST_ERR_SUCCESS) {
+                psa_write(msg.handle, 0, out_vec[0].base, out_vec[0].len);
```
no check if write failed?

moranpeker replied:
no need. ```psa_write``` is a void func

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,248 @@
+            set_caller_id(msg);
+            status = initial_attest_get_token_size(in_vec, 1, out_vec, 1);
+            if (status == PSA_ATTEST_ERR_SUCCESS) {
+                psa_write(msg.handle, 0, out_vec[0].base, out_vec[0].len);
```
no check if write failed?

moranpeker replied:
psa_write is void function

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,248 @@
+                                                     &public_key_data_length);
+
+            if (status == PSA_SUCCESS) {
+                psa_write(msg.handle, 0, public_key_data, public_key_data_length);
```
no check if write failed?

moranpeker replied:
psa_write is void function

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,248 @@
+                psa_write(msg.handle, 0, public_key_data, public_key_data_length);
+            }
+
+            psa_write(msg.handle, 1,
```
no check if write failed?

moranpeker replied:
psa_write is void function

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,248 @@
+                break;
+            }
+            if (msg.in_size[1] != 0) {
+                key_data = calloc(1, msg.in_size[1]);
```
where is key_data being freed?

moranpeker replied:
fixed

alekshex reviewed (commented):
please see latest comments

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/psa_inject_attestation_key_impl.h:
```diff
@@ -0,0 +1,50 @@
+* limitations under the License.
+*/
+
+/***************************************************************************/
```
do we need this comment here?

moranpeker replied:
removed

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attestation_bootloader_data.c:
```diff
@@ -0,0 +1,49 @@
+    0x31, 0x32
+};
+
+uint8_t impl_id_data[TEMP_IMPL_ID_DATA_SIZE] = {TEMP_IMPL_ID_DATA};
```
legal eof character?

moranpeker replied:
fixed

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_iat_claims_loader.c:
```diff
@@ -0,0 +1,228 @@
+    psa_status_t crypto_ret;
+    psa_hash_operation_t hash_handle;
+
+    crypto_ret = psa_crypto_init();
```
do we keep init here?

moranpeker replied:
removed.

alekshex commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto_keys.c:
```diff
@@ -0,0 +1,195 @@
+    /* Copy the x-coordinate of public key to the buffer */
+    if (initial_attestation_public_x_key_size != 0)
+    {
+        key_src = key_src + ONE_BYTE;
```
can you please write short comment explaining why +1

moranpeker replied:
Its already described 12 line before

Commit:
Add newline at end of file

Commit:
Remove calling to psa_crypto_init

- no need to call psa_crypto_init before crypto operations
we assume init already called before. if not an error will return.

Commit:
Free allocated buffers

Commit:
Using goto exit in attest_public_key_sha256 function

Commit:
General fixes of spelling, newline, un needed code

-Correct spelling errors
-Add newline at end of file
-Remove un needed code

Commit:
astyle fix

Commit:
Remove calling to "psa_crypto_init()"

No need to call psa_crypto_init before crypto operations
we assume init already called before. if not an error will return.

Commit:
Add newline at end of file and remove un needed commant

kjbracey commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/tfm_impl/t_cose/src/t_cose_util.c:
```diff
@@ -0,0 +1,189 @@
+     * probability of collision with this ID is very low and the same
+     * as for collision between any two key IDs of any sort.
+     */
+    uint8_t defined_short_circuit_kid[] = {
```
This should be `static const`, presumably?

moranpeker replied:
This file is taken from TFM attestation code.
Should we review this code too?

kjbracey replied:
If it's upstream code, may be best to just leave it if making changes causes admin hassle. This doesn't cause a RAM size impact, just teeny code impact.

kjbracey commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attestation_bootloader_data.c:
```diff
@@ -0,0 +1,49 @@
+    0x31, 0x32
+};
+
+uint8_t impl_id_data[TEMP_IMPL_ID_DATA_SIZE] = {TEMP_IMPL_ID_DATA};
```
Should be `const`

moranpeker replied:
fixed

kjbracey commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_iat_claims_loader.c:
```diff
@@ -0,0 +1,225 @@
+{
+    enum tfm_plat_err_t status;
+    uint8_t *p_dst;
+    uint8_t p_src[ATTEST_PUB_KEY_SHA_256_SIZE];
```
There's no need for this extra buffer - nothing's stopping you outputting straight into the final destination. (The thing you're calling could length limit if necessary, but you don't need it to anyway).

moranpeker replied:
Right, Fixed

kjbracey commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attestation_bootloader_data.h:
```diff
@@ -0,0 +1,58 @@
+** between the report and verification */
+#define TEMP_IMPL_ID_DATA_SIZE (32u)
+
+#define TEMP_IMPL_ID_DATA  0xD0, 0xD1, 0xD2, 0xD3, 0xD4, 0xD5, 0xD6, 0xD7, \
```
Is there a reason why this is in the public header file, rather than in the C file?

moranpeker replied:
You are right, fixed

kjbracey commented on components/TARGET_PSA/services/attestation/COMPONENT_SPE/psa_attestation_partition.c:
```diff
@@ -0,0 +1,259 @@
+
+static void psa_attest_get_token(void)
+{
+    psa_msg_t msg = { 0 };
```
This is just blowing code space, isn't it? Or is it needed to handle a silent `psa_get` failure?

Generally a bad idea to initialise variables to dummy data, as it can waste ROM space (particularly if a structure), and it stops the compiler or valgrind's memcheck from spotting paths where you don't fill in the variables with their real data.

orenc17 replied:
@kjbracey-arm this is the default way to declare the psa_msg_t type
in most cases it will be zeroed by psa_get

i don't see an issue here

kjbracey replied:
> in most cases it will be zeroed by psa_get

I think that neatly articulates the issue. The current code poses a number of questions: Is it an output parameter for `psa_get`, or is it a read-modify-write parameter? Can we affect the behaviour of `psa_get` by having different stuff in the message on entry? Under what circumstances does `psa_get` not write to it?

By not initialising it, that's a clear statement that it's an output-only parameter, and the compiler and/or valgrind can check that (a) `psa_get` doesn't read it and (b) `psa_get` always writes the fields that you rely on.

With the initialisation, the code is just fuzzy about who's supposed to be writing to it, and who reads from it.

And on top of that, this is actually making the code bigger. So that's two negative strikes, and no advantage I can see.

(The output-only API argument would suggest it might be even better to have the API be `psa_msg_t psa_get(psa_signal_t signum)`, but that can be less efficient if the compiler fails to perform return value optimisation, so it's not totally clear-cut. It would be idiomatic C++11 though).

kjbracey reviewed (commented):
Just some drive-by code style and optimisation nits

Commit:
Remove header files from inc into services/attestation folder

Commit:
Remove incorrect free place

Commit:
Update TFM attestation code to latest version

The latest and greatest version is:
c43181daf54f69f53de58593a50dd6a9c233eecd

- Add new header file and implement tfm_memcpy (used in
attetstaion_core c file)

Commit:
Update astyleignore

Ignore attesttaion.h that taken from TFM

Commit:
Remove TEMP_IMPL_ID_DATA and TEMP_IMPL_ID_DATA_SIZE

remove from public header into attestation_bootloader_data.c and add
const to impl_id_data

Commit:
Update astyleignore

correct file path

Commit:
Remove extra buffers used

outputting straight into the final destination

Commit:
Add free to allocated buffer while error

Commit:
Fix attestation main to check PSA target before includes

0xc0170 commented on components/TARGET_PSA/services/attestation/qcbor/test/qcbor_encode_tests.c:
```diff
@@ -0,0 +1,2007 @@
+ Copyright (c) 2018-2019, Laurence Lundblade.
+ All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
```
Any non apache files should be documented . See https://os.mbed.com/docs/mbed-os/v5.11/contributing/license.html

This needs addition to our Readme.md and LICENSE.md file with license description in the attestation folder or whichever folder its root for this software (I can see Readme there, that is good ! Just add LICENSE.md with the license and add reference in our root Readme).

moranpeker replied:
Done.

Commit:
Documented any non apache files

Commit:
Correct psa-crypto include

Commit:
Styling

Commit:
Remove includes from mbedtls

Commit:
Use less stack

orenc17 reviewed (approved):
Approving this with a big comment for maintainers, this code is not tested nor it enabled on TFM

Commit:
astyle fix

Commit:
Prevent compiler warning, converting psa_key_policy_t to const in call to psa_set_key_policy

Commit:
remove unsupported psa target

Commit:
fix void pointer increasing with bytes

Commit:
astyle fix

tamasban commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -0,0 +1,277 @@
+/*
+* Copyright (c) 2018-2019 ARM Limited. All rights reserved.
+*
+* SPDX-License-Identifier: Apache-2.0
```
Pls check whether is it OK from legal point of view to replace the license header in this file. It seems that this file is derived from TF_M repo ./lib/t_cose/src/t_cose_psa_crypto.c

0xc0170 replied:
@moranpeker Please review

Commit:
Add SW_SIGNER_ID data to boot status

Commit:
initialized  psa_hash_operation_t to zeros before calling hash_setup

-initialized  psa_hash_operation_t to zeros
-fix TF-M attestation code - increase t_cose_crypto_hash bytes to
handle max psa_hash_operation_t size

Commit:
change license to file that mostly based on TF-M implementation

from Apache-2.0 to BSD-3-Clause

0xc0170 commented on components/TARGET_PSA/services/attestation/COMPONENT_PSA_SRV_IMPL/attest_crypto.c:
```diff
@@ -1,20 +1,12 @@
+ *
+ * Copyright (c) 2018-2019, Laurence Lundblade. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
```
just checking, this attestation is already in Readme covered as non apache file and license file present in the attestation folder?

0xc0170 replied:
Answer: yes, I've just checked

Commit:
add casting to increased void pointer
